### PR TITLE
add 'update' method in main loop for generic internal operations.

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

The PR proposes a new abstract method in the `Agent` class, the `update()` method. This method is called at every iteration of the main loop of the agent (that is, the one in the method `run_main_loop()`). The purpose of the `update()` method is to handle generic operations to update the state of the agent. 

Notice how it differs from the `act()` method, that handles the outputs of the agent, and the `react()` method, that handles the inputs of the agent.
Moreover, it's worth noticing that, at the moment, this is just a logical separation of operations, rather than an increase in expressiveness. There's no practical difference between the `react` and the `update` method.

## Fixes

This PR is somehow related to #295. That's because some operations that we used to accomplish by allocating threads now can be completed inside this method.

Among other changes, the PR brings also small fixes to the source code.

## Types of changes

What types of changes does your code introduce to agents-tac?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [X] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [X] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

- It could be possible to generalize this method by including a `task_queue` that consumes small tasks, so instead of requiring the programmer to implement the `update()` method, he could just enqueue some tasks in the `task_queue`, delegating its execution to the framework. That implies the need for defining the right abstractions, error handling etc.
- Another remark is that, while before this change the cleanup job of the `TransactionManager` run every `_cleanup_locked_transactions_task_timeout` seconds, now it runs at the same rate of the main loop. That might not be the desired behaviour. The replication of the same behaviour requires some efforts to the developer. So it's worth thinking about a way to make his life easier.

### CODE REVIEW

File: tac/agents/v1/agent.py [hunk 1]
```diff
@@ -134,3 +135,10 @@ def react(self) -> None:
 
         :return: None
         """
+
+    @abstractmethod
+    def update(self) -> None:
+        """Update the current state of the agent.
```
Comment (DavidMinarsch) [inline tac/agents/v1/agent.py]: Note, this is ok for now, but going forward we need to introduce more explicit language around what is the agent state vs what is game specific 'AgentState'. The former is a much wider domain than the latter.

File: tac/agents/v1/base/game_instance.py [hunk 2]
```diff
@@ -362,13 +361,13 @@ def state_after_locks(self, is_seller: bool) -> AgentState:
...
 
-    def generate_proposal(self, cfp_services: Dict[str, Union[bool, List]], is_seller: bool) -> List[Description]:
+    def generate_proposal(self, cfp_services: Dict[str, Union[bool, List]], is_seller: bool) -> Optional[List[Description]]:
         """
         Wrap the function which generates proposals from a seller or buyer.
 
         If there are locks as seller, it applies them.
 
-        :param query: the query associated with the cfp.
+        :param cfp_services: the query associated with the cfp.
```
Comment (DavidMinarsch) [inline tac/agents/v1/base/game_instance.py]: change to: the services associated with the cfp

**Comment by DavidMinarsch** [comment]
I agree with your comment on the `TaskQueue`, something we should keep in mind for future architecture changes.

File: tac/agents/v1/agent.py (hunk 1 above)
Comment (marcofavorito) [reply tac/agents/v1/agent.py]: Agreed!

File: tac/agents/v1/base/game_instance.py (hunk 2 above)
Comment (marcofavorito) [reply tac/agents/v1/base/game_instance.py]: Done!

**Pull request merged**